src/sentry/utils/event_tracker.py

```diff
@@ -0,0 +1,55 @@
+import logging
+from enum import IntEnum
+
+from sentry.utils.hashlib import md5_text
+
+
+class EventStageStatus(IntEnum):
+    REDIS_PUT = 1
+    """
+    redis_put
+
+    save_event_started
+
+    save_event_finished
+
+    snuba_topic_put
+
+    commit_log_topic_put
+
+    ppf_topic_put
+
+    post_process_started
+
+    post_process_finished / the same as redis_deleted
+    """
+
+
+logger = logging.getLogger("EventTracker")
+
+
+def is_sampled_to_track(event_id: str, sample_rate) -> bool:
+    """
+    Normalize the integer to a float in the range [0, 1)
+    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
+    and provide all or nothing logging.
+    """
+    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
+    if hash_float < sample_rate:
```

Comment:
As discussed on Slack: the max value of the numerator in the hash_float calculation is the same as the denominator, so hash_float has a range of [0, 1]. I'd recommend using `<=` for the sample rate comparison - in the infinitesimally small case of exactly 0 or 1, I'd rather incorrectly log at sample rate `0` than miss an event at sample rate `1.0`.

```suggestion
    Normalize the integer to a float in the range [0, 1]
    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
    and provide all or nothing logging.
    """
    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
    if hash_float <= sample_rate:
```

Reply:
Alternatively to avoid the edge case (but I don't think it really matters)

```suggestion
    Normalize the integer to a float in the range [0, 1]
    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
    and provide all or nothing logging.
    """
    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
    if sample_rate > 0 && hash_float <= sample_rate:
```

Reply:
Please do not rely on sampling = 0.0 to let any event pass. That would also be your killswitch. 
If anything you should do one of theese two:
- checked whether sampling is 0.0 and quit immediately before hashing the event id.
- have another option to serve as a killswitch so that if it is off you do not do anything inside the method including hashing.

Reply:
added if sample rate ==0 return